File: p260m/single_number3.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var singleNumber = function (nums) {
  const xor = nums.reduce((s, num) => s ^ num, 0);
  const lastsetbit = xor & -xor; // extract the last set bit
  const ans = [0, 0];
  for (let n of nums) {
    if (n & lastsetbit) ans[0] ^= n;
    else ans[1] ^= n;
  }
  return ans;
};
// TEST
[
  [
    [1, 2, 1, 3, 2, 5],
    [3, 5],
  ],
  [
    [-1, 0],
    [-1, 0],
  ],
  [
    [0, 1],
    [0, 1],
  ],
  [
    [1, 5, 5, 1, 98, 98, 23, 58, -1, -1],
    [23, 58],
  ],
].forEach(([nums, expected]) => {
  const actual = singleNumber(nums).sort();
  console.log('Two single nums in', nums, '->', actual);
  console.assert(actual.toString() === expected.toString());
});

File: p260m/single_number3.py
from typing import List
from functools import reduce


class Solution:
    def singleNumber(self, nums: List[int]) -> List[int]:
        x = reduce(lambda a, b: a ^ b, nums, 0)
        # another way to extrac the last set bit, x & (x-1) remove the last bit
        # then ^x will restore it only
        last_set_bit = x & (x - 1) ^ x
        num1 = reduce(
            lambda a, b: a ^ b, filter(lambda n: n & last_set_bit, nums)
        )
        return num1, x ^ num1


# TESTS
for nums, expected in [
    [
        [1, 2, 1, 3, 2, 5],
        [3, 5],
    ],
    [[-1, 0], [-1, 0]],
    [[0, 1], [0, 1]],
    [
        [1, 5, 5, 1, 98, 98, 23, 58, -1, -1],
        [23, 58],
    ],
]:
    sol = Solution()
    actual = sol.singleNumber(nums)
    print("Two single nums in", nums, "->", actual)
    assert sorted(actual) == expected
